File: 144-Binary-Tree-Preorder-Traversal.cpp
#include <vector>
#include <stack>

using std::vector;
using std::stack;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// class Solution {
// public:
//     void visit(TreeNode* root, vector<int>& result) {
//         if (root == NULL) { return; }
//         result.push_back(root->val);
//         visit(root->left, result);
//         visit(root->right, result);

//     }
//     vector<int> preorderTraversal(TreeNode* root) {
//         vector<int> result;
//         visit(root, result);
//         return result;
//     }
// };

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        if (root == NULL) {
            return result;
        }

        stack<TreeNode*> stk;
        stk.push(root);
        while (!stk.empty())
        {
            TreeNode* tn = stk.top();
            stk.pop();
            result.push_back(tn->val);
            if (tn->right != NULL) {
                stk.push(tn->right);
            }

            if (tn->left != NULL) {
                stk.push(tn->left);
            }
        }
        
        return result;
    }
};

File: 144-Binary-Tree-Preorder-Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


//  func visit(tn *TreeNode, result *[]int) {
// 	if (tn == nil) {
// 		return
// 	}
// 	*result = append(*result, tn.Val)
// 	visit(tn.Left, result)
// 	visit(tn.Right, result)
// }

// func preorderTraversal(root *TreeNode) []int {
//     var result []int
// 	visit(root, &result)
// 	return result
// }

func preorderTraversal(root *TreeNode) []int {
	var result []int
	
	if root == nil {
		return result
	}

	var stk []*TreeNode
	stk = append(stk, root)
	for len(stk) > 0 {
		tn := stk[len(stk) - 1]
		stk = stk[:len(stk) - 1]
		result = append(result, tn.Val)
		if (tn.Right != nil) {
			stk = append(stk, tn.Right)
		}
		if (tn.Left != nil) {
			stk = append(stk, tn.Left)
		}
	}

	return result
}